### PR TITLE
Return results from jsonpatch.apply; call onRemoteChange outside observe. #85

### PR DESCRIPTION
**Pull request by joozek78**

### CODE REVIEW

**Comment by joozek78** [comment]
depends on starcounter-jack/JSON-patch#110

File: src/puppet.js [hunk 1]
```diff
@@ -471,6 +471,18 @@
 
   Puppet.prototype = Object.create(EventDispatcher.prototype); //inherit EventTarget API from EventDispatcher
 
+  Puppet.prototype.dispatchErrorEvent = function(error) {
```
Comment (tomalec) [inline src/puppet.js]: Do we need it in public API?

File: src/puppet.js [hunk 2]
```diff
@@ -560,44 +572,52 @@
...
+      } else {
+        throw error;
       }
     }
-    else {
-      this.jsonpatch.apply(tree, sequence);
+
+    var that = this;
+    sequence.forEach(function (patch) {
+      if (patch.path === "") {
```
Comment (tomalec) [inline src/puppet.js]: @warpech do we need this check in non-debug mode?

File: src/puppet.js (hunk 1 above)
Comment (joozek78) [inline src/puppet.js]: No, it's just my inexperience with JS showed its ugly head :) I'll move it

File: src/puppet.js [hunk 3]
```diff
@@ -622,40 +642,18 @@
...
 
     if (this.onPatchReceived) {
         this.onPatchReceived(data, url, method);
     }
 
-    if (!this.observer) {
-      return; //ignore remote change if we are not watching anymore
-    }
+    // apply only if we're still watching
+    if (this.observer) {
```
Comment (tomalec) [inline src/puppet.js]: Personally I prefer "return fast" in simple cases, as then I do not have to read entire method, and search for applicable `else`

File: test/spec/OnLocalRemoteChangeSpec.js [hunk 4]
```diff
@@ -60,11 +58,40 @@ describe("OnLocal/RemoteChange", function() {
...
                 "op": "replace",
                 "path": "/hello",
                 "value": "onRemoteChange callback"
             }]);
+            expect(mostRecentCall.args[1]).toEqual([{removed: 'world'}]);
+            done();
+        }, 100); // for FF
+    });
+
+    it("should fire patch-applied event for applied patches", function(done) {
```
Comment (tomalec) [inline test/spec/OnLocalRemoteChangeSpec.js]: Could you also add a test  for `should NOT fire patch-applied event for received, and queued patches`?

File: src/puppet.js (hunk 3 above)
Comment (joozek78) [inline src/puppet.js]: I just find it clearer and more intuitive, but I don't have to stick to it.

Comment (tomalec) [inline src/puppet.js]: For me it's less intuitive, as to preserve democracy, you can ask Warpech ;)

**Pull request merged**